README.md
# easy-webmention-js

![](https://img.shields.io/github/release/ggirelli/easy-webmention-js.svg?style=flat) ![](https://img.shields.io/github/release-date/ggirelli/easy-webmention-js.svg?style=flat) ![](https://img.shields.io/github/languages/code-size/ggirelli/easy-webmention-js.svg?style=flat)  
![](https://img.shields.io/github/license/ggirelli/easy-webmention-js.svg?style=flat) ![](https://img.shields.io/github/watchers/ggirelli/easy-webmention-js.svg?label=Watch&style=social) ![](https://img.shields.io/github/stars/ggirelli/easy-webmention-js.svg?style=social)

A simple JavaScript library to serve [Webmentions](https://www.w3.org/TR/webmention/) via [Webmention.io](https://Webmention.io)

**\[ [Live Demo](https://ggirelli.info/blog/2021/08/08/webmentions#webmentions-wrap) ]**

## Requirements

This script requires [jQuery](https://jquery.com/) v3.3.1+ and [Bootstrap](https://getbootstrap.com/) v5.0.2+.

## Usage

Please, add the following in the `<head />` of your pages, after importing Bootstrap and jQuery:
```html
<script type="text/javascript" src="https://ggirelli.github.io/easy-webmention-js/src/easy-webmention.min.js"></script>
```

Then, place the following where you would like your webmentions to appear:

```html
<!-- WEBMENTIONS -->
<div id="webmentions-wrap" class="border-top pt-2"></div>
<script type="text/javascript">
    $(function () {
        ew_init({
                      "tag" : "{{ site.webmentions.tag }}",       // Webmention.io username
                   "target" : "{{ site.url }}{{ page.url }}",     // This page's URL
            "read_more_uri" : "{{ site.webmentions.help_page }}", // URL to help page (optional)
                  "wrap_id" : "#webmentions-wrap"                 // ID of element where to add webmentions
        });
    });
</script>
```


CHANGELOG.MD
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).



## Unreleased

## [1.0.1] - 2021-08-22
### Fixed
- "View on..." link now correctly positioned on bottom-right corner.

### Added
- Separate cards for short elements (e.g., mentions, reposts, likes,...).

## [1.0.0] - 2021-08-15

[Unreleased]: https://github.com/ggirelli/easy-webmention-js
[1.0.1]: https://github.com/ggirelli/easy-webmention-js/releases/tag/v1.0.1
[1.0.0]: https://github.com/ggirelli/easy-webmention-js/releases/tag/v1.0.0


src/easy-webmention.js
/*! easy-webmention-js - v1.0.1 - 2021-08-22
 * https://github.com/ggirelli/easy-webmention-js
 * Copyright (c) 2021
 * License: MIT
 * Author: Gabriele Girelli https://ggirelli.info
 */

plural_labels = {
    "like": "likes",
    "mention": "mentions",
    "reply": "replies",
    "bookmark": "bookmarks",
    "repost": "re-posts"
};

verbal_labels = {
    "like-of": "Liked",
    "mention-of": "Mentioned",
    "bookmark-of": "Bookmarked",
    "repost-of": "Re-posted"
};

function ew_append_element(m_property, index, item, target) {
    card_body = $("<div />")
        .attr("id", "ew-webmention-" + m_property + "-" + index)
        .attr("title", Date(Date.parse(item.published)))
        .addClass("row card-body");

    if ( item.author.name == "" ) { item.author.name = "Anonymous"; }
    if ( item.author.url == "" ) { item.author.url = "#" }
    if ( item.author.photo == "" ) { item.author.photo = "t.ly/a0z5"; }
    author_avatar = $("<img />")
            .attr("src", item.author.photo)
            .attr("alt", item.author.name + "-avatar")
            .addClass("ew-author-avatar");
    card_body.append($("<div />")
        .addClass("col-3 col-md-1 pe-0")
        .append(author_avatar));

    item_text = "View";
    if ( item.url.startsWith("https://twitter.com/") ) { item_text += " on Twitter"; }
    if ( item.url.startsWith("https://github.com/"
        ) || item.url.startsWith("https://www.github.com/") ) {
        item_text += " on GitHub"; }

    author_link = $("<a />")
            .addClass("ew-author-link text-decoration-none")
            .attr("href", item.author.url).attr("target", "_new")
            .text(item.author.name);
    item_link = $("<a />")
            .addClass("ew-webmention-link")
            .addClass("text-decoration-none fs-6 lh-1")
            .attr("href", item.url).attr("target", "_new")
            .text(item_text);

    item_content = $("<div />")
        .addClass("col-9 col-md-11 pb-2")
        .append(author_link);
    if ( item.content != null ) {
        item_content.append($("<div />").text(item.content.text));
    } else {
        item_content.append($("<div />").text("Content could not be retrieved."
            ).css({"font-style":"italic"}));
    }
    item_content.append($("<div />")
        .addClass("position-absolute bottom-0 end-0 p-1 pe-2")
        .append(item_link));
    card_body.append(item_content);

    $(target)
        .append($("<div />")
            .addClass("card mb-3")
            .append($("<a />").attr("id", "ew-webmention-" + index))
            .append(card_body));
}

function ew_append_short_element(m_property, index, item, target) {
    card_body = $("<div />")
        .attr("id", "ew-webmention-" + m_property + "-" + index)
        .attr("title", new Date(Date.parse(item["wm-received"])))
        .addClass("row card-body py-1");

    silos_link = $("<a />").attr("href", item.url).addClass("text-decoration-none");
    if ( item.url.startsWith("https://twitter.com/") ) { silos_link.text("Twitter"); }
    if ( item.url.startsWith("https://github.com/"
        ) || item.url.startsWith("https://www.github.com/") ) {
        silos_link.text("GitHub"); }

    item_date = new Date(Date.parse(item["wm-received"])).toLocaleDateString(
        "en-US", { year: 'numeric', month: 'short', day: 'numeric',
                   hour: '2-digit', minute: '2-digit' });

    author_avatar = $("<img />")
        .attr("src", item.author.photo)
        .attr("alt", item.author.name + "-avatar")
        .addClass("ew-author-avatar");

    card_body.append($("<table />").append($("<tr />")
        .append($("<td />")
            .css({"width":"2.5em"})
            .append($("<div />")
                .addClass("d-block overflow-hidden")
                .css({"width":"2.5em", "height":"2.5em"})
                .append(author_avatar .addClass("d-inline"))))
        .append($("<td />")
            .append($("<span />")
                .addClass("ms-2")
                .text(verbal_labels[m_property] + " by "))
            .append($("<a />")
                .addClass("text-decoration-none")
                .attr("href", item.author.url)
                .text(item.author.name))
            .append($("<span />").text(" on "))
            .append(silos_link))
        .append($("<td />")
            .append($("<small />")
                .addClass("float-end d-table-cell")
                .css({"line-height":"2em", "font-size":".75em"})
                .text(item_date)))
    ));

    $(target)
        .append($("<div />")
            .addClass("card mb-3 bg-transparent border-0")
            .append($("<a />").attr("id", "ew-webmention-" + index))
            .append(card_body));
}

function ew_load_property(ew_settings, m_property, short=false) {
    $.ajaxSetup({async: false});
    $.get("https://webmention.io/api/mentions.jf2?target=" +
        ew_settings.target + "&wm-property=" + m_property, {}, function(response) {
        if ( 0 != response.children.length ) {
            $.each(response.children, function(index, item) {
                if ( short ) {
                    ew_append_short_element(m_property, index, item, $(ew_settings.wrap_id));
                } else {
                    ew_append_element(m_property, index, item, $(ew_settings.wrap_id));
                }
            });
        }
    });
}

function ew_prep_read_more_button(ew_settings) {
    return($("<a/>")
        .addClass("ps-2 float-end")
        .attr("id", "ew-read-more")
        .attr("target", "_new")
        .attr("href", ew_settings.read_more_uri)
        .attr("data-bs-toggle", "tooltip" )
        .attr("data-bs-placement", "top")
        .attr("title", "Read more on webmentions")
        .html("<i class='fas fa-question-circle'></i>"))
};

function ew_prep_add_button(ew_settings) {
    return($("<a/>")
        .addClass("ps-2 float-end")
        .attr("id", "ew-add-webmention")
        .attr("href", "#")
        .attr("data-bs-toggle", "tooltip" )
        .attr("data-bs-placement", "top")
        .attr("title", "Manually add a webmention")
        .html("<i class='fas fa-plus-circle'></i>")
        .click(function(e) {
            e.preventDefault();
            if ( $(".ew-add-form").hasClass("d-none") ) {
                $(".ew-add-form").removeClass("d-none").slideDown();
            } else {
                $(".ew-add-form").slideUp(400, function() {
                    $(this).addClass("d-none")
                });
            }
        })
    )
}

function ew_prep_add_form(ew_settings) {
    return($("<form />")
        .attr("action", "https://webmention.io/" + ew_settings.tag + "/webmention")
        .attr("method", "post")
        .attr("target", "_new")
        .addClass("ew-add-form")
        .append($("<small />")
            .addClass("d-block mb-1 text-secondary lh-1 fs-6")
            .html("To manually add a webmention, type the address to a page containing " +
                  "a link to this article, then press enter.<br />This will open the " +
                  "webmention.io response page. Refresh the page if the status is pending."))
        .append($("<input />")
            .attr("type", "hidden")
            .attr("name", "target")
            .attr("value", ew_settings.target))
        .append($("<input />")
            .addClass("form-control")
            .attr("type", "text")
            .attr("name", "source")
            .attr("placeholder", "..."))
    )
}

function ew_prep_header_count(count) {
    return($("<p />").addClass("fs-5 header-block")
        .append($("<span/>").text(count + " webmentions")));
}

function ew_prep_header_buttons(ew_settings) {
    if ( ew_settings.read_more_uri == undefined ) {
        return($("<small/>")
                .append(ew_prep_add_button(ew_settings)));
    } else {
        return($("<small/>")
                .append(ew_prep_read_more_button(ew_settings))
                .append(ew_prep_add_button(ew_settings)))
    }
}

function ew_init(ew_settings) {
    $.get("https://webmention.io/api/count?target=" + ew_settings.target,
        {}, function(response, status) {
            ew_wrap = $(ew_settings.wrap_id);

            // HEADER
            ew_title_element = ew_prep_header_count(response.count);
            ew_title_element.append(ew_prep_header_buttons(ew_settings))
            ew_title_element.append(ew_prep_add_form(ew_settings));
            ew_wrap.append(ew_title_element);
            new bootstrap.Tooltip($("#ew-add-webmention")).enable();
            new bootstrap.Tooltip($("#ew-read-more")).enable();
            $(".ew-add-form").slideUp(0, function() {
                $(this).addClass("d-none");
            })

            // COUNTS AND ITEMS
            if ( 0 < response.count ) {
                $.each(response.type, function(label, value) {
                    item_header = $("<h6 />").addClass("mb-3");
                    if ( value > 1 && plural_labels[label] != undefined) {
                        item_header.text(value + " " + plural_labels[label]);
                    } else {
                        item_header.text(value + " " + label);
                    }
                    ew_wrap.append(item_header);
                    
                    if ( "reply" == label && response.type.reply > 0 ) {
                        ew_load_property(ew_settings, "in-reply-to");
                    }
                    if ( "mention" == label && response.type.mention > 0 ) {
                        ew_load_property(ew_settings, "mention-of");
                    }
                    if ( "repost" == label && response.type.repost > 0 ) {
                        ew_load_property(ew_settings, "repost-of", true);
                    }
                    if ( "like" == label && response.type.like > 0 ) {
                        ew_load_property(ew_settings, "like-of", true);
                    }
                });
            }
        }
    );
}
